flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/state/rocksdb/iterator/RocksMultiStateKeysIterator.java

```diff
@@ -0,0 +1,180 @@
+    private final int keyGroupPrefixBytes;
+    private final byte[] namespaceBytes;
+    private final DataInputDeserializer byteArrayDataInputView;
+
+    private final byte[][] iteratorKeys;
+    private K previousKey;
+    private K nextKey;
+
+    public RocksMultiStateKeysIterator(
+            List<RocksIteratorWrapper> iterators,
+            List<String> states,
+            @Nonnull TypeSerializer<K> keySerializer,
+            int keyGroupPrefixBytes,
+            List<Boolean> ambiguousKeyPossibles,
+            @Nonnull byte[] namespaceBytes) {
+        this.iterators = iterators;
+        this.states = states;
+        this.keySerializer = keySerializer;
+        this.ambiguousKeyPossibles = ambiguousKeyPossibles;
+        this.keyGroupPrefixBytes = keyGroupPrefixBytes;
+        this.namespaceBytes = namespaceBytes;
+        this.byteArrayDataInputView = new DataInputDeserializer();
+        this.iteratorKeys = new byte[iterators.size()][];
+        Arrays.fill(iteratorKeys, null);
+        this.previousKey = null;
+        this.nextKey = null;
+    }
+
+    @Override
+    public boolean hasNext() {
+        try {
+            while (nextKey == null && hasDataToProcess()) {
+                pullKeysFromIterators();
+                K smallestIteratorKey = calculateSmallestKeyFromLocalData();
+                if (smallestIteratorKey != null) {
+                    previousKey = smallestIteratorKey;
+                    nextKey = smallestIteratorKey;
+                }
+            }
+        } catch (Exception e) {
+            throw new FlinkRuntimeException(
+                    "Failed to access states [" + String.join(",", states) + "]", e);
+        }
+        return nextKey != null;
+    }
+
+    private boolean hasDataToProcess() {
+        boolean result = iterators.stream().anyMatch(RocksIteratorWrapper::isValid);
+        if (!result) {
+            for (int i = 0; i < iterators.size(); ++i) {
+                if (iteratorKeys[i] != null) {
+                    result = true;
+                    break;
+                }
+            }
+        }
+        return result;
+    }
+
+    private void pullKeysFromIterators() {
+        for (int i = 0; i < iterators.size(); ++i) {
+            RocksIteratorWrapper iterator = iterators.get(i);
+            if (iteratorKeys[i] == null && iterator.isValid()) {
+                iteratorKeys[i] = iterator.key();
+                iterator.next();
+            }
+        }
+    }
+
+    @Nullable
+    private K calculateSmallestKeyFromLocalData() throws IOException {
+        int smallestIteratorKeyIndex = -1;
+        byte[] smallestIteratorKey = null;
+        for (int i = 0; i < iteratorKeys.length; ++i) {
+            byte[] iteratorKey = iteratorKeys[i];
+            if (iteratorKey != null
+                    && (smallestIteratorKey == null
+                            || Arrays.compare(iteratorKey, smallestIteratorKey) < 0)) {
```

Comment:
Minor optimization: if `Arrays.compare(iteratorKey, smallestIteratorKey) == 0` here, just remove the corresponding `iteratorKeys[i]`? Avoid some compare below.

Reply:
That's a nice catch, started testing it...

Reply:
Well, the blind removal was not enough but with a bit more complex logic we've eliminated quite some compares. I don't see any huge performance gain but it can be as we increase the number of iterators.

Reply:
Okey, let's keep it.